### PR TITLE
MGMT-10734: Add host states and transitions for reclaim

### PR DESCRIPTION
**Pull request by carbonin**
Adds the statemachine transitions and new states in order to implement
the following diagram:

```
                   -----------------------
                   | HostStatusInstalled |
                   -----------------------
                             |
                      User unbinds agent
                             |
     Kube API and no BMH     |   REST API, Kube API with BMH, or other machine management
          -------------------+-------------------
          |                                     |
          V                                     V
-----------------------------            ----------------------------
| TransitionTypeReclaimHost |            | TransitionTypeUnbindHost |
-----------------------------            ----------------------------
            |                                               |
            V                                               |
  ------------------------                                  |
  | HostStatusReclaiming |                                  |
  ------------------------                                  |
                |                                           |
       Agent posts step reply                               |
                |                                           |
       ---------+----------------------------------         |
       |                                          |         |
    Success                                    Failure      |
       |                                          |         |
       V                                          |         |
-------------------------------------             |         |
| HostStatusReclaimingPendingReboot |------       |         |
-------------------------------------     |       |         |
                |                         |       |         |
  ------------------------------       Timeout    |         |
  | TransitionTypeRegisterHost |          |       |         |
  ------------------------------          |       |         |
                |                         |       |         |
                V                         V       V         V
--------------------------------  ----------------------------------------
| HostStatusDiscoveringUnbound |  | HostStatusUnbindingPendingUserAction |
--------------------------------  ----------------------------------------
              ^                                       |
              |       ------------------------------  |
              --------| TransitionTypeRegisterHost |---
                      ------------------------------

```

Specifically I'm looking for input on the names of these states in this patch. This doesn't actually make any of the transitions happen yet.

## List all the issues related to this PR

Related to the enhancement in https://github.com/openshift/assisted-service/pull/3849
Part 1 of https://issues.redhat.com/browse/MGMT-10734

The next PR will add the code to push the hosts through these new states when required.

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

This code is just adding states and transitions, no functional changes were added.

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @avishayt 
/cc @filanov 
/cc @nmagnezi 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

### CODE REVIEW

**Comment by avishayt** [comment]
1. What is HostStatusReclaimingPendingReboot?  Do we move there just before the reboot to understand the progress?
2. For Kube API with BMH, do we want to skip the generic method as in the diagram, or keep the same flow regardless and use BMH as a fallback?

**Comment by carbonin** [comment]
>What is HostStatusReclaimingPendingReboot? Do we move there just before the reboot to understand the progress?

Effectively, yeah.
I wanted to split the two so that we could get a step response after things like the image downloads and writing out the boot entry. Then `HostStatusReclaimingPendingReboot` will tell the agent to just reboot and we'll progress when the agent registers.

File: internal/host/pool_host_statemachine.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ func NewPoolHostStateMachine(sm stateswitch.StateMachine, th *transitionHandler)
 			stateswitch.State(models.HostStatusKnownUnbound),
 			stateswitch.State(models.HostStatusUnbinding),
 			stateswitch.State(models.HostStatusUnbindingPendingUserAction),
+			stateswitch.State(models.HostStatusReclaimingPendingReboot),
```
Comment (avishayt) [inline internal/host/pool_host_statemachine.go:30]: PendingReboot sounds like the user needs to do it
```suggestion
			stateswitch.State(models.HostStatusReclaimingRebooting),
```

**Comment by avishayt** [comment]
> I'm not sure I understand. If we have the BMH I'd prefer we use that and not try to start an agent on the node.
> We could probably improve the way it's done currently, but I think it's out of scope for this feature.

1. Originally when we were thinking about this feature we discussed all kinds of shortcuts like kexec, but now that we need to reboot and this isn't necessarily more reliable than BMH, we can skip it for the BMH flow.
2. We need to think about how to detect the BMH flow.  Currently, the only controller that deals with BMH is BMAC and I wouldn't want that logic to creep into the rest of Assisted.

**Comment by carbonin** [comment]
> now that we need to reboot and this isn't necessarily more reliable than BMH, we can skip it for the BMH flow.

Yeah, I think BMH will handle more situations than this will so it's better to default to using that if it's present.

> We need to think about how to detect the BMH flow. Currently, the only controller that deals with BMH is BMAC and I wouldn't want that logic to creep into the rest of Assisted.

This is one of the reasons that we can assume this feature will only be available in the kube API.
Given that it's only available in the kube API I don't see why the agent controller wouldn't be able to search for a BMH on unbind by sharing the matching logic with BMAC (or by just using a label selector).
Hopefully that doesn't qualify as creeping into the rest of assisted.

**Comment by carbonin** [comment]
Yeah, right now the plan is just to indicate implicitly (linked to a bmh) that we shouldn't reclaim the agent. We could add something explicit as well, but I'm not sure it's relevant for this PR, right?

**Comment by avishayt** [comment]
> Yeah, right now the plan is just to indicate implicitly (linked to a bmh) that we shouldn't reclaim the agent. We could add something explicit as well, but I'm not sure it's relevant for this PR, right?

Not relevant for this PR, just noting it.

**Pull request merged**